service/history/replication/executable_history_task.go

```diff
@@ -299,3 +305,134 @@ func (e *ExecutableHistoryTask) getDeserializedEvents() (_ []*historypb.HistoryE
+	incomingHistoryTask, isHistoryTask := incomingTask.(*ExecutableHistoryTask)
+	if !isHistoryTask {
+		return nil, serviceerror.NewInvalidArgument("Unsupported Batch type")
+	}
+
+	if err := e.checkWorkflowKey(incomingHistoryTask.WorkflowKey); err != nil {
+		return nil, err
+	}
+
+	if err := e.checkVersionHistoryItem(incomingHistoryTask.versionHistoryItems); err != nil {
+		return nil, err
+	}
+
+	if err := e.checkBaseExecutionInfo(incomingHistoryTask.baseExecutionInfo); err != nil {
+		return nil, err
+	}
+
+	events, newRunEvents, err := incomingHistoryTask.getDeserializedEvents()
+	if err != nil {
+		return nil, err
+	}
+
+	if err = e.checkEvents(events, newRunEvents); err != nil {
+		return nil, err
+	}
+
+	return incomingHistoryTask, nil
+}
+
+// checkVersionHistoryItem will check if incoming tasks Version history is on the same branch as the current one
+func (e *ExecutableHistoryTask) checkVersionHistoryItem(incomingHistoryItems []*historyspb.VersionHistoryItem) error {
+	if versionhistory.IsVersionItemsInSameBranch(e.versionHistoryItems, incomingHistoryItems) {
+		return nil
+	}
+	return serviceerror.NewInvalidArgument("version history does not match")
+}
+
+func (e *ExecutableHistoryTask) checkWorkflowKey(incomingWorkflowKey definition.WorkflowKey) error {
+	if !(e.WorkflowKey.GetWorkflowID() == incomingWorkflowKey.GetWorkflowID() &&
+		e.WorkflowKey.GetNamespaceID() == incomingWorkflowKey.GetNamespaceID() &&
+		e.WorkflowKey.GetRunID() == incomingWorkflowKey.GetRunID()) {
+		return serviceerror.NewInvalidArgument("workflow key does not match")
+	}
+	return nil
+}
+
+func (e *ExecutableHistoryTask) checkBaseExecutionInfo(incomingTaskExecutionInfo *workflowpb.BaseExecutionInfo) error {
+	if e.baseExecutionInfo != nil && incomingTaskExecutionInfo != nil && !e.baseExecutionInfo.Equal(incomingTaskExecutionInfo) {
+		return serviceerror.NewInvalidArgument("base execution info not match")
+	}
+
+	return nil
+}
+
+func (e *ExecutableHistoryTask) checkEvents(incomingEvents []*historypb.HistoryEvent, incomingNewRunEvents []*historypb.HistoryEvent) error {
+	if len(incomingEvents) == 0 {
+		return serviceerror.NewInvalidArgument("incoming task is empty")
+	}
+	currentEvents, currentNewRunEvents, err := e.getDeserializedEvents()
+	if err != nil {
+		return err
+	}
+
+	if currentNewRunEvents != nil {
+		return serviceerror.NewInvalidArgument("Current Task is expected to be the last event of a workflow")
+	}
+	currentLastEvent := currentEvents[len(currentEvents)-1]
+	incomingFirstEvent := incomingEvents[0]
+	if currentLastEvent.Version != incomingFirstEvent.Version {
+		return serviceerror.NewInvalidArgument("events version mismatch")
+	}
+	if currentLastEvent.EventId+1 != incomingFirstEvent.EventId {
+		return serviceerror.NewInvalidArgument("events id is not consecutive")
+	}
+
+	return nil
+}
+
+func (e *ExecutableHistoryTask) CanBatch() bool {
+	return e.wantToBatch
```

Comment:
lock?

Reply:
if this function assumes that lock is already acquired, then rename to `canBatchLocked`